Tickets/dm 12563

mfisherlevine opened the pull request.

mfisherlevine committed:
Copy all of obs_comCam over

mfisherlevine committed:
Rename all comCam/ComCam to auxTel/AuxTel

mfisherlevine committed:
Add .gitignore

mfisherlevine committed:
Change how some values are derived during ingest.

Path patterns are different for the TS3 datasets so change
how we calculate the run number and a few other things. Remove
'raft' entry as there is no raft. Change 'ccd' to zero as there
is only one of them. We no longer seem to have jobIds so remove
this (obs_comCam never uses them from what I can remember anyway)

Put in a safety-check that the serial in the head matches that in
the path as I don't trust the info all that much. Also, a simple
assert here is no good, as it gets swallowed by a try block, whereas
putting in a raise with a message will show you why that file was
skipped.

mfisherlevine committed:
Fix ignored cached file

mfisherlevine committed:
Change ccd type to int from text.

There was a big gap in time between making this change and commiting
it, and I can't quite remember why it was necessary, but it was, and
the error you get if you change this is quite strange, so this is
something to watch out for.

mfisherlevine committed:
Add filters

Add an alias for the empty filter, and the other filters found
in the data we have. This is a living list, and needs to be kept
up-to-date as more data is taken with different filters with
different names.

mfisherlevine committed:
Clean up the template paths a little

Arguably there perhaps shouldn't be any mention of the CCD in these
templates, but I think that the CCDs having names/numbers is assumed in
many places, so it just seemed a lot simpler for this package, given that
we're trying to keep the creation as simple as possible, to just keep
it in, and always return 0 for the CCD number.

mfisherlevine committed:
First commit in changing the amplifier geom

Rip out the e2v definition, and the constructs which tile the CCDs into
the raft. Next commit will fix up the actual geometry changes, as this
is not currently assembling correctly in its current state.

mfisherlevine committed:
Change full amplifier bbox to the right size

Contains a follow-up ticket to change how this is used to make
it consistent with the rest of the bboxes in the yaml cameras.
See DM-13098 for the implementation ticket.

SimonKrughoff commented on config/bias.py:
```diff
@@ -0,0 +1,2 @@
+config.isr.doLinearize = False
```
Here and elsewhere in the `config` directory, it would be nice to have a comment on why defaults are being overridden.

SimonKrughoff commented on config/ingest.py:
```diff
@@ -0,0 +1,45 @@
+    # Note that we DO ingest a run number, but it comes from the path and not the header,
+    # so an entry here generates a warning. It is picked up in the ingest.py override
+    # of getInfo() and registers correctly like this with the line below left commented out.
+    # 'run': 'RUNNUM'
```
I think this can be left out completely.  We can mention in the skeleton that a run number is nice to have and should be filled in either from the header or from a translation function.

mfisherlevine replied:
Sure, sounds good, I've removed. Just to be clear though, this is not the same thing as a visit number, and I'm not sure that for normal operations that a run number is actually something that we'll want/need.

SimonKrughoff replied:
We will definitely want to specify observations taken for different purposes.  This, in principle, could be encoded in a run number.

mfisherlevine replied:
Oh yes, for sure. I'd assume that between `OBSTYPE, IMGTYPE` and `OBJECT` we'd have that covered, and I wanted to remove `RUNNUM` in the meantime as it's confusing as it's not currently being used quite like that in the lab data we have, but it could well end up augmenting the keywords I just mentioned on the sky, and might want to be added back in in the future.

SimonKrughoff commented on config/ingest.py:
```diff
@@ -0,0 +1,45 @@
+config.parse.hdu = 0
+
+config.register.columns = {
+    # int is *probably* OK for run, but I'm not sure, and it's useful to be the same as for obs_comCam
```
Does this need an issue in Jira.  It seems like it's probably a ducktyping problem where the pipelines assume `ccd` is an `int` somewhere in the internals.
  
It sounds like it, at least, needs a mention in the obs cookbook.

mfisherlevine replied:
Hmmmm. Again, run numbers shouldn't be confused with visit numbers, nor ccds. I do think that specifying a datatype for them would be advisable; having everything being freeform certainly doesn't make downstream code cleaner, and I don't see why, in the new paradigm, we shouldn't enforce `ints` for at least `visits` and `ccds`, IMHO at least.

mfisherlevine replied:
For now I've just removed the comment to clean it up though and left it as `text` as obs_comCam _needs_ it to be text, as their run numbers have a "D" on the end of about half of them, and we shouldn't have our code fall over when it meets that.

SimonKrughoff commented on config/ingestCalibs.py:
```diff
@@ -0,0 +1,18 @@
+config.parse.retarget(AuxTelCalibsParseTask)
+
+config.register.columns = {'filter': 'text',
+                           'ccd': 'text',
```
I find it concerning that `ccd` needs to be `int` if the file is a normal exposure, but not if it is a calib.  Should we be consistent, just to beat down the cognitive dissonance?

mfisherlevine replied:
Yes! That's not cool at all, thanks for spotting that.

mfisherlevine replied:
Also, in an ideal world there _would be no ccd_ number/text in this camera as it's a single chip, but that would have made more work than the time allotted allowed for.

SimonKrughoff commented on config/ingest.py:
```diff
@@ -0,0 +1,45 @@
+    'field': 'text',
+    'wavelength': 'int',
+}
+config.register.visit = list(config.register.columns.keys())
```
Do you really need all the keys in order to define a unique visit?  Correct me if I'm wrong, but that should be the set of keys that defines a unique visit and it seems like requiring them all is a bit much.  I could be wrong, though.

mfisherlevine replied:
Oh boy, good spot, and this has made me look at this more, leading to the following:
1) No, I certainly don't.
2) I really don't know why this is like this, I just cargo-culted it from obs_comCam.
3) but I'm now not sure which keys to pick to trim this down.
*but* most importantly:
4) this is _not_ `config.register.unique`, this is `config.register.`_visit_, and I honestly don't know why there is no `unique` being done here at all. Why do some packages specify unique, and others don't?

mfisherlevine replied:
Anyway, action taken: leave as-is, but with a comment saying that the columns registered can be culled if the registry becomes too large, and a value for `config.register.unique` which I'm pretty sure for this camera is just `'visit'`.

SimonKrughoff commented on policy/auxTelMapper.yaml:
```diff
@@ -0,0 +1,411 @@
+defects: ../description/defects
```
There is quite a lot of duplication between this and what is already defined in `obs_base`.  I know it's a pain, but it would be really nice to remove as much of that duplication now as possible.  The cleaner the repo is to start, the easier it will be to keep cruft out.

mfisherlevine replied:
For now, I've just taken them all out, as they're not needed at all (as we don't have any on-sky data yet!). Have also added comments on what each item _is_.

SimonKrughoff commented on policy/camera.yaml:
```diff
@@ -0,0 +1,188 @@
+    rawParallelPrescanBBox  : [[0,    1], [0,         0]]  # pixels digitised before first parallel transfer
+    rawParallelOverscanBBox : [[0, 2000], [509,      48]]  # parallel overscan
+
+    saturation : 262144                  # saturation level, DN XXX Should this be in electrons?
```
I think this depends on whether we mean numerical saturation or full well.  I was assuming that it mean numerical saturation and that full well was slightly higher than numerical saturation.

mfisherlevine replied:
This should actually be used to mean full-well, and _not_ numerical saturation, and it's actually the other way around to what you might (do) think, and the reason is ~interesting: the full-well needs to be digitizable in order to apply cross-talk corrections correctly.

However, without knowing the actual full-well values, using the numerical saturation is a decent placeholder.

More importantly though, what would you like done with this line of code? Just leave as-is?

SimonKrughoff commented on policy/auxTelMapper.yaml:
```diff
@@ -0,0 +1,411 @@
+    tables: raw
+    template: raw/%(ccd)s/%(ccd)s-%(visit)06d.fits[%(channel)d]
+  postISRCCD:
+    template: 'postISRCCD/postISRCCD_v%(visit)d_%(ccd)s_f%(filter)s.fits'
```
This is the only template that is quoted.  I think it would be nice to be consistent and remove the quotes if this survives the culling of duplicated material.

mfisherlevine replied:
I have no idea why it was like this, nor why the others weren't. They are all the same now.

SimonKrughoff commented on python/lsst/obs/auxTel/auxTelMapper.py:
```diff
@@ -0,0 +1,338 @@
+    #
+    # We need to standardize, but have no legal way to call std_raw.  The butler should do this for us.
+    #
+    atm = AuxTelMapper()
```
I'd prefer a slightly more descriptive variable name.  `atm` sounds like "atmosphere".

SimonKrughoff commented on python/lsst/obs/auxTel/auxTelMapper.py:
```diff
@@ -0,0 +1,338 @@
+        for d in (self.mappings, self.exposures):
+            d['raw'] = d['_raw']
+
+        # self.filterIdMap = {}           # where is this used?  Generating objIds??
```
If this is unused, it should go away.

SimonKrughoff commented on python/lsst/obs/auxTel/auxTelMapper.py:
```diff
@@ -0,0 +1,338 @@
+        @param dataId (dict) Data identifier with visit
+        """
+        visit = dataId['visit']
+        return int(visit)
```
Why does this need to be coerced into an `int`?

mfisherlevine replied:
I honestly don't know, this is totally cargo-culting. It's probably safe to remove it in this case, so I guess I will... 😬

mfisherlevine replied:
Nope, just worked it out. The docstring says that this is "the 64-bit (long) identifier for a CCD exposure." so I think this is just to guarantee that, so I'll leave this in for now as I think that the bit-length is important here.

SimonKrughoff commented on python/lsst/obs/auxTel/auxTelMapper.py:
```diff
@@ -0,0 +1,338 @@
+        return dataId["ccd"]
+
+    def _computeCcdExposureId(self, dataId):
+        """Compute the 64-bit (long) identifier for a CCD exposure.
```
Probably all methods should have parameter definitions in their docstrings.

At this point, the docstring format also changes from sphinx style to doxygen style.  I think it would be good to use sphinx throughout.

mfisherlevine replied:
Oh yeah, gross, I totally agree!

SimonKrughoff commented on python/lsst/obs/auxTel/auxTelMapper.py:
```diff
@@ -0,0 +1,338 @@
+        return self.map__raw_visitInfo(*args, **kwargs)
+
+    def bypass_raw_visitInfo(self, datasetType, pythonType, location, dataId):
+        if False:
```
It doesn't feel right to have this here.  I think this block should be made into a Jira issue this `if False` block removed.

mfisherlevine replied:
I have just removed it completely for now, I think that's OK. Shout if you want something different to happen here.

SimonKrughoff commented on python/lsst/obs/auxTel/auxTelMapper.py:
```diff
@@ -0,0 +1,338 @@
+        return self._standardizeExposure(self.exposures['raw_amp'], item, dataId,
+                                         trimmed=False, setVisitInfo=False)
+
+    def X_validate(self, dataId):
```
I don't understand all these "not currently live" methods.  Is it that you are planning on using these in the future but do not at the moment?  If that's the case, I think we can remove them here and leave them on a branch or something.

mfisherlevine replied:
I believe this was an attempt, during the birth of the previous package to mitigate the snowballing cargo-culting, and by "turning them off", and seeing which were actually necessary when it came to runtime. Given that this is the second package which has inherited them with them being re-enabled (albeit without processing sky-data) I've just gone and deleted them for now. We can always re-cargo-cult them from other packages, so I don't think there is a need for a branch here (plus there is always the git history)

SimonKrughoff commented on python/lsst/obs/auxTel/ingest.py:
```diff
@@ -0,0 +1,129 @@
+            The recorded wavelength as an int
+        """
+        raw_wl = md.get("MONOWL")
+        wl = int(round(raw_wl))
```
Doesn't `round` return an `int`?  It does in python3, but maybe not in python2.

mfisherlevine replied:
You're right, but it doesn't do much harm to ensure that it's an int so that it works for those still in py2 land does it. Do you think this needs removing?

SimonKrughoff commented on python/lsst/obs/auxTel/ingest.py:
```diff
@@ -0,0 +1,129 @@
+##############################################################################################################
+
+
+class AuxTelCalibsParseTask(CalibsParseTask):
```
It would be great to have sphinx docstrings in this class too.  I understand that they are `_` methods, but we do that with `_` methods elsewhere.

SimonKrughoff commented on ups/obs_auxTel.table:
```diff
@@ -0,0 +1,14 @@
+setupRequired(afw)
```
I think only modules that are actually imported need to be in this table file.  For example, `skypix` is not used anywhere in this module.  I suspect there are others.

SimonKrughoff commented on tests/SConscript:
```diff
@@ -0,0 +1,3 @@
+# -*- python -*-
```
`obs_base` provides some tests that can be automatically run if you provide some mapping information.  It seems like at least those tests should be run.

SimonKrughoff reviewed (commented):
Generally looks great.  There are some consistency issues and possible cargo culting that we can remove.  See comments for specifics.

I would like to see at least the minimal tests running.

mfisherlevine committed:
Comment on override of defaults

mfisherlevine committed:
Remove comment on the run number ingest key

mfisherlevine committed:
Remove comment about run datatype for cleanliness

mfisherlevine committed:
Add config.register.unique

Also, add comment on registering all keys

mfisherlevine committed:
Change ccd type to int for calibs in keeping with regular visits

mfisherlevine committed:
Remove most datasets from yaml mapper definition

Most of these either do, or should, have defaults in obs_base.
Remove everything not needed for all the pre-on-sky part of the pipeline
processing for now, and add back as required in order to reduce bloat,
cargo-culting and code duplication.

mfisherlevine committed:
Add brief description for each item

mfisherlevine committed:
Remove errant quotes from one template

mfisherlevine committed:
Rename variable to avoid confusing name

mfisherlevine committed:
Remove commented line

mfisherlevine committed:
Improve and update all docstrings to new format

mfisherlevine committed:
Remove unused methods

Deleted all methods prefixed with X_ as these are not currently
being used. Also removed the consequently unusued imports.

mfisherlevine committed:
Improve docstrings

mfisherlevine committed:
Thin out UPS table file
